## **Data Cleaning: Stack Overflow Annual Developer Survey Results**

In [1]:
# import all the necessary libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

# Ignore Python Warnings
import warnings
warnings.filterwarnings("ignore")

In [2]:
# import the data

dataset = pd.read_csv(r"E:\Big Datasets\Stack Overflow Data\survey_results_public.csv")

df = dataset.copy()

df.head()

ResponseId                      MainBranch                 Age  \
0           1  I am a developer by profession  Under 18 years old   
1           2  I am a developer by profession     35-44 years old   
2           3  I am a developer by profession     45-54 years old   
3           4           I am learning to code     18-24 years old   
4           5  I am a developer by profession     18-24 years old   

            Employment RemoteWork   Check  \
0  Employed, full-time     Remote  Apples   
1  Employed, full-time     Remote  Apples   
2  Employed, full-time     Remote  Apples   
3   Student, full-time        NaN  Apples   
4   Student, full-time        NaN  Apples   

                                    CodingActivities  \
0                                              Hobby   
1  Hobby;Contribute to open-source projects;Other...   
2  Hobby;Contribute to open-source projects;Other...   
3                                                NaN   
4                                                NaN   

                                             EdLevel  \
0                          Primary/elementary school   
1       Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
2    Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
3  Some college/university study without earning ...   
4  Secondary school (e.g. American high school, G...   

                                           LearnCode  \
0                             Books / Physical media   
1  Books / Physical media;Colleague;On the job tr...   
2  Books / Physical media;Colleague;On the job tr...   
3  Other online resources (e.g., videos, blogs, f...   
4  Other online resources (e.g., videos, blogs, f...   

                                     LearnCodeOnline  ... JobSatPoints_6  \
0                                                NaN  ...            NaN   
1  Technical documentation;Blogs;Books;Written Tu...  ...            0.0   
2  Technical documentation;Blogs;Books;Written Tu...  ...            NaN   
3  Stack Overflow;How-to videos;Interactive tutorial  ...            NaN   
4  Technical documentation;Blogs;Written Tutorial...  ...            NaN   

  JobSatPoints_7 JobSatPoints_8 JobSatPoints_9 JobSatPoints_10  \
0            NaN            NaN            NaN             NaN   
1            0.0            0.0            0.0             0.0   
2            NaN            NaN            NaN             NaN   
3            NaN            NaN            NaN             NaN   
4            NaN            NaN            NaN             NaN   

  JobSatPoints_11           SurveyLength SurveyEase ConvertedCompYearly JobSat  
0             NaN                    NaN        NaN                 NaN    NaN  
1             0.0                    NaN        NaN                 NaN    NaN  
2             NaN  Appropriate in length       Easy                 NaN    NaN  
3             NaN               Too long       Easy                 NaN    NaN  
4             NaN              Too short       Easy                 NaN    NaN  

[5 rows x 114 columns]

In [3]:
df.shape

(65437, 114)

##### Dataset consists of 65,437 rows and 114 columns

In [10]:
# Ensure all rows are shown
pd.set_option('display.max_rows', None)

summary_df = pd.DataFrame({
    'Non-Null Count': df.notnull().sum(),
    'Null Count': df.isnull().sum(),
    'Dtype': df.dtypes
})
print(summary_df)

                                Non-Null Count  Null Count    Dtype
ResponseId                               65437           0    int64
MainBranch                               65437           0   object
Age                                      65437           0   object
Employment                               65437           0   object
RemoteWork                               54806       10631   object
Check                                    65437           0   object
CodingActivities                         54466       10971   object
EdLevel                                  60784        4653   object
LearnCode                                60488        4949   object
LearnCodeOnline                          49237       16200   object
TechDoc                                  40897       24540   object
YearsCode                                59869        5568   object
YearsCodePro                             51610       13827   object
DevType                                  59445  

In [5]:
# Import the data dictionary

data_dictionary = pd.read_csv(r"E:\Big Datasets\Stack Overflow Data\survey_results_schema.csv")

data_dictionary.head()

qid       qname                                           question  \
0    QID2  MainBranch  Which of the following options best describes ...   
1  QID127         Age                                 What is your age?*   
2  QID296  Employment  Which of the following best describes your cur...   
3  QID308  RemoteWork  Which best describes your current work situation?   
4  QID341       Check  Just checking to make sure you are paying atte...   

  force_resp type selector  
0       True   MC     SAVR  
1       True   MC     SAVR  
2       True   MC     MAVR  
3      False   MC     SAVR  
4       True   MC     SAVR

In [6]:
df.dtypes.value_counts()

object     100
float64     13
int64        1
Name: count, dtype: int64

In [11]:
df.memory_usage(deep=True).sum()

np.int64(472368432)

In [12]:
472368432 / 1024 / 1024

450.4856414794922

In [14]:
# Checking duplicates

df.duplicated().sum()

np.int64(0)

In [ ]:
# All the columns
for col in df.columns:
    print(col)

ResponseId
MainBranch
Age
Employment
RemoteWork
Check
CodingActivities
EdLevel
LearnCode
LearnCodeOnline
TechDoc
YearsCode
YearsCodePro
DevType
OrgSize
PurchaseInfluence
BuyNewTool
BuildvsBuy
TechEndorse
Country
Currency
CompTotal
LanguageHaveWorkedWith
LanguageWantToWorkWith
LanguageAdmired
DatabaseHaveWorkedWith
DatabaseWantToWorkWith
DatabaseAdmired
PlatformHaveWorkedWith
PlatformWantToWorkWith
PlatformAdmired
WebframeHaveWorkedWith
WebframeWantToWorkWith
WebframeAdmired
EmbeddedHaveWorkedWith
EmbeddedWantToWorkWith
EmbeddedAdmired
MiscTechHaveWorkedWith
MiscTechWantToWorkWith
MiscTechAdmired
ToolsTechHaveWorkedWith
ToolsTechWantToWorkWith
ToolsTechAdmired
NEWCollabToolsHaveWorkedWith
NEWCollabToolsWantToWorkWith
NEWCollabToolsAdmired
OpSysPersonal use
OpSysProfessional use
OfficeStackAsyncHaveWorkedWith
OfficeStackAsyncWantToWorkWith
OfficeStackAsyncAdmired
OfficeStackSyncHaveWorkedWith
OfficeStackSyncWantToWorkWith
OfficeStackSyncAdmired
AISearchDevHaveWorkedWith
AISearchDevWantTo

##### **Observation:**
- The dataset has 65,437 observations
- Their are total 114 variables:
  - ```int```: 1
  - ```float```:13
  - ```object```:100
- The dataset utilizes 472,368,432 of memory
  - Roughly 450 mega-bytes of memory
- Their are some columns with missing values
  - ```ResponseId```, ```MainBranch```, ```Age```, ```Employment```, ```RemoteWork``` and ```Check``` columns have no null values
  - Values are missing most likely because the users chose to skip/ignore the questions
- Many columns contain inconsisent characters and potentially incorrect data types
- The dataset do not contain any duplicated values

In [ ]:
# Dropping Unnecessary columns

df = df.drop(columns=['OpSysPersonal use','OpSysProfessional use', 'ResponseId'])

In [22]:
# Replace Missing Values with "Unanswered"

df = df.fillna("Unanswered")

In [23]:
df.isnull().sum()

MainBranch                        0
Age                               0
Employment                        0
RemoteWork                        0
Check                             0
CodingActivities                  0
EdLevel                           0
LearnCode                         0
LearnCodeOnline                   0
TechDoc                           0
YearsCode                         0
YearsCodePro                      0
DevType                           0
OrgSize                           0
PurchaseInfluence                 0
BuyNewTool                        0
BuildvsBuy                        0
TechEndorse                       0
Country                           0
Currency                          0
CompTotal                         0
LanguageHaveWorkedWith            0
LanguageWantToWorkWith            0
LanguageAdmired                   0
DatabaseHaveWorkedWith            0
DatabaseWantToWorkWith            0
DatabaseAdmired                   0
PlatformHaveWorkedWith      

## **Column-wise Inspection**

In [46]:
df['MainBranch'].unique()

array(['I am a developer by profession', 'I am learning to code',
       'I code primarily as a hobby',
       'I am not primarily a developer, but I write code sometimes as part of my work/studies',
       'I used to be a developer by profession, but no longer am'],
      dtype=object)